Урок  13 Создал тесты для методов ping и create booking

ismailtester opened the pull request.

ismailtester committed:
Создал тесты для методов ping и create booking

dmitryvshivtsev commented on tests/test_create_booking.py:
```diff
@@ -0,0 +1,18 @@
+    jsonschema.validate(response_json, CREATED_BOOKING_SCHEMA)
+
+
+
```
Может добавишь ещё негативный кейс с использованием pytest.raises?
https://docs-python.ru/packages/frejmvork-pytest/funktsija-pytest-raises/

dmitryvshivtsev replied:
Отправь, например, пустой словарь и убедись, что выбрасывается исключение при ошибке сервера

dmitryvshivtsev replied:
а лучше ещё пару каких-то сценариев придумай
негативных или позитивных

ismailtester committed:
Добавлены автотесты для метода create_booking:

Позитивные тесты:
1. Создание бронирования только с обязательными полями

Негативные тесты:
2. Невалидные типы данных (xfail, ожидается 400) — по моему мнению, должна приходить 400-я ошибка, а не 500, поэтому тест отмечен как xfail
3. Пустое тело запроса (xfail, ожидается 400) — аналогично, ожидаю 400 вместо 500, поэтому xfail
4. Недоступность сервера (Exception) — юнит
5. Таймаут соединения (requests.Timeout) — юнит
6. Ошибочный HTTP метод (405) — юнит
7. Неправильный URL (404) — юнит
8. Ошибка сервера (500) — юнит
9. Успешный, но неожиданный статус 201 вместо 200 — юнит

dmitryvshivtsev commented on tests/test_create_booking.py:
```diff
@@ -0,0 +1,118 @@
+    jsonschema.validate(response_json, CREATED_BOOKING_SCHEMA)
+
+
+@pytest.mark.xfail(reason="Known bug: API returns 500 instead of 400 on invalid input")
+@allure.feature("Test Create Booking")
+@allure.story("Create booking with only required fields and invalid data")
+@pytest.mark.parametrize("field,value", [
+    ("firstname", 123123),
+    ("lastname", 123123),
+    ("totalprice", "cheap"),
+    ("depositpaid", "not paid"),
+    ("bookingdates", "12.04.2025")
+])
+def test_create_booking_with_invalid_data(api_client, generate_random_booking_data, field, value):
+    payload = copy.deepcopy(generate_random_booking_data)
+    payload[field] = value
+    with pytest.raises(requests.HTTPError) as exc_info:
+        api_client.create_booking(payload)
+    status_code = exc_info.value.response.status_code
+    assert status_code == 400, f"Expected 400 Bad Request, but got {status_code}"
```
В данном случае сервер действительно отдаёт 500. Просто считаем, что в этом случае сервер так себя ведет и это норма. Можно не накидывать xfail. Но вообще ход хороший

dmitryvshivtsev commented on tests/test_create_booking.py:
```diff
@@ -0,0 +1,118 @@
+    status_code = exc_info.value.response.status_code
+    assert status_code == 400, f"Expected 400 Bad Request, but got {status_code}"
+
+@allure.feature("Test Create Booking")
+@allure.story("Test server unavailability")
+def test_create_booking_server_unavailable(api_client, mocker, generate_random_booking_data):
+    payload = generate_random_booking_data
+    mocker.patch.object(api_client.session, "post", side_effect=Exception("Server unavailable"))
+
+    with pytest.raises(Exception, match="Server unavailable"):
+        api_client.create_booking(payload)
```
Этот и следующий лучше не мокать в этих случаях. Если ты мокаешь, то проверяешь, что твой клиент и его метод будут себя определенным образом вести в случае такого ответа от сервера. А нам здесь нужно проверить реальное поведение сервера. Для этого мы и отправляем неверные данные, а потом проверяем, что пришло 500 (что не совсем корректно, но это проблема со стороны песочницы)
Можно эти тесты вынести в отдельный файл, который именно проверяет работу твоего клиента.
